language-plutus-core/src/Language/PlutusCore/CBOR.hs

```diff
@@ -272,3 +285,88 @@ instance Serialise UnknownDynamicBuiltinNameError
+
+Serialising the unit annotation takes up space: () is converted to the
+CBOR `null` value, which is encoded as the byte 0xF6.  In typical
+examples these account for 30% or more of the bytes in a serialised
+PLC program.  We don't actually need to serialise unit annotations
+since we know where they're going to appear when we're deserialising,
+and we know what the value has to be.  The `InvisibleUnit` type below
+has instances which takes care of this for us: if we have an
+`InvisibleUnit`-annotated program `prog` then `serialise prog` will
+serialise a program omitting the annotations, and `deserialise` (with
+an appropriate type ascription) will give us back an
+`InvisibleUnit`-annotated program.
+
+We usually deal with ()-annotated ASTs, so the annotations have to be
+converted to and from `InvisibleUnit` if we wish to save space.  The
+obvious way to do this is to use `InvisibleUnit <$ ...` and
+`() <$ ...`, but these have the disadvantage that they have to traverse the
+entire AST and visit every annotation, adding an extra cost which may
+be undesirable when deserialising things on-chain.  However,
+`InvisibleUnit` has the same underlying representation as `()`, and
+we can exploit this using Data.Coerce.coerce to convert entire ASTs
+with no run-time overhead.
+-}
+
+newtype InvisibleUnit = InvisibleUnit ()
+
+instance Serialise InvisibleUnit where
+    encode = mempty
+    decode = pure (InvisibleUnit ())
+
+{- Note [Serialising Scripts]
+
+At first sight, all we need to do to serialise a script without unit
+annotations appearing in the CBOR is to coerce from `()` to
+`InvisibleUnit` and then apply `serialise` as normal.  However, this
+isn't sufficient for the ledger code. There are a number of places in
+Ledger.Scripts and elsewhere where hashes of objects (`Tx`, for
+example) are calculated by serialsing the entire object and
+calculating a hash, and these objects can contain Scripts themselves.
+Serialisation is achieved using the Generic instance of `Serialise`,
+which will just use `encode` on everything, including unit
+annotations.  There are two ways to overcome this.  Firstly, we could
+write explicit `Serialise` instances for everything.  This would be
+more space-efficient than the Generic encoding, but would introduce a
+lot of extra code.  The other way is to make `Ledger.Scripts.Script`
+a special instance of `Serialise` which performs the coercions for us.
+The functions below facilitate this, and are used to implement a suitable
+instance for `Script` in `Ledger.Scripts`.
+-}
+
+encodePLC :: (Closed uni, uni `Everywhere` Serialise) => Program TyName Name uni () -> Encoding
+encodePLC (p :: Program TyName Name uni ()) = encode (coerce p :: Program TyName Name uni InvisibleUnit)
+
+decodePLC :: (Closed uni, uni `Everywhere` Serialise) => Decoder s (Program TyName Name uni ())
+decodePLC = do
+      p :: Program TyName Name uni InvisibleUnit <- decode
+      return $ (coerce p :: Program TyName Name uni ())
+
```

Comment:
I feel that these should really be called something like `encodePLCOmmittingUnitAnnotations` and `decodePLCRestoringUnitAnnotations`, but that's a bit unwieldy.